resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb

```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+module OpenTelemetry
+  module Resource
+    module Detectors
+      # Deployment contains detect class method for determining deployment resource attributes
+      module Deployment
+        extend self
+
+        def detect
+          resource_attributes = {}
+          deployment_environment = rails_deployment_environment || rack_env
```

Comment:
Personally speaking I think it would be OK to rely on only using environment variables instead of framework specific helpers. 

So in this case I think it would be fine to use something like `ENV['RACK_ENV'] || ENV['RAILS_ENV'] || ENV['APP_ENV']`

Reply:
The issue is that the frameworks have a default environment of "development" in case ENV is missing.
To reflect that in the resource, I will need to write `ENV['RACK_ENV'] || ENV['RAILS_ENV'] || ENV['APP_ENV'] || "development"`, which will work for rails and sinatra but will incorrectly report "development" environment if these frameworks are not being used.

Reply:
Totally understand. Will you be using sinatra integration in a separate PR?

Reply:
I wasn't planing to add Sinatra support for this PR.
Tried to extract the env like this:
```rb 
          def sinatra_env
            # https://github.com/sinatra/sinatra/blob/e69b6b9dee7165d3a583fc8a6af10ceee1ea687d/lib/sinatra/base.rb#L1801
            ::Sinatra::Base::environment.to_s if defined?(::Sinatra::Base::environment)
          end  
```

Which works fine if the `require 'sinatra'` is placed before the SDK is initialized.
If user configures the SDK and only then require 'sinatra', then this will fail. In addition, it looks like the sinatra instrumentation does not work in this case (we should probably open an issue and debug it as an other PR).

Do you have any suggestion on how to extract the value in this case?

Reply:
Thank you for giving it a try and thank you for your patience as at this point I've written more comments that actual code in this PR. 

I think it is reasonable to add development dependencies for testing. Require the dependencies as part of the test case and it will make them available to the system under test. You can do this as part of Bundle.require or explicitly requiring it in the test case.

That combined with mocking and `unstub_cost` helpers should be enough to do the trick.

Reply:
Thank you too for the high-quality review and help on these topics.
did some experimenting with sinatra. I was able to get it to work with the help of @YanivD . If you can take another look at the updated code it could be great :)

As documented in the code, this resource detector will not be able to extract the default enviroment "development" if sinatra is used and it is required after resource detector is run (resource is immutable and there is no way to know if it will be rrequired in the future at the time of resource creation).

This is the relevant part of the code for reference:
```rb
        def sinatra_env
          # https://github.com/sinatra/sinatra/blob/e69b6b9dee7165d3a583fc8a6af10ceee1ea687d/lib/sinatra/base.rb#L1801
          # cases:
          #
          # 1. if sinatra is "require"d before the detector, then we return the value from the library
          # this case will return the default "development" fallback if not env variable is set which is good.
          #
          # 2. if sinatra is "require"d after the detector, then:
          # 2.1 if user is setting environment via 'APP_ENV' or 'RACK_ENV' then those value will be picked up and reported
          # 2.2 else, the sinatra environment will fallback to "development", but detector will return nil for it.
          # this issue is not covered, as when detector initialize the immutable resource, it has no way
          # of knowing if "sinatra" will be required later or not.
          (::Sinatra::Base.environment.to_s if defined?(::Sinatra::Base.environment)) || ENV['APP_ENV']
        end
```